### ISSUE TITLE
trying to extend arrayjoin

### ISSUE THREAD

**Issue by drrudolph** [libvips/libvips#1344]
Hi,
first of all thanks for the work on the library, it seems to be a good fit for what we are trying to do.

I am trying to montage several thousand images into a empty large image (possibly up to 40GPixel). The images are not in a rectilinear grid and I already have a given array of coordinates for alignment.

What I am planning to do is write an "vips_arraymontage" function (based on vips_arrayjoin), handing over the coordinate array as a parameter and then search the coordinates in vips_arrayjoin_gen() if any tiles overlap the output region.

My questions:
1) Does that sound feasible ?
2) I can't seem to find a way to give a VipsRect array as a parameter in vips_arraymontage_class_init() (VIPS_TYPE_ARRAY_RECT ?). What is a good way to do that ?
(I'm new to gobjects, so this is probably something obvious...)

I am talking about this part:

	VIPS_ARG_BOXED( class, "coords", -1,
		_( "coordinates" ),
		_( "Array of input image coordinates as VipsRects" ),
		VIPS_ARGUMENT_REQUIRED_INPUT,
		G_STRUCT_OFFSET( VipsArraymontage, coords ),
		VIPS_TYPE_ ????);

3) Is it bad memory-wise to call functions with the same variable as input and output
e.g. vips_affine(image, &image,...) ? This seems to be avoided in all the library calls which use a
vips_object_local_array()

Thanks

**Comment by jcupitt** [comment]
Hi @drrudolph, yes, that sounds reasonable. 

You could also look at `composite`: it does exactly what you want, though with Porter-Duff blending. The one in libvips 8.8 has been revised so that it will perform well with thousands of non-overlapping inputs. If you give it plain RGB images, it will treat them as opaque. It'll do greyscale, and 8/16/32 bit pixels as well.

I would make a test program in Python and benchmark it.

One performance problem you might hit is that `composite` will do a simple linear search for every output rect to find the set of inputs which touch it. This is fine for 10s of inputs, but might be slow for many 1000s. I would replace the linear search with a map: split the output coordinate space into (for example) a 16 x 16 grid, then for each grid point, note the list of inputs which overlap it. Then for each output rect, test against the list of inputs on the set of grid squares which touch that rect.

Another possible problem is file descriptor limits. Windows has a hard limit of 2000 files open at once. Linux allows any number (I use 65536), but you will need to configure your system for it.

Rather than passing an array of rect, I would pass separate int arrays for x and y, as `composite` does. You'll find this is much more convenient to interface to languages like Python.

You can use `vips_affine(image, &image, ...)`, but be careful about reference leaks. I usually write something like:

```C
VIpsImage *image = ...;
VipsImage *x;

if (vips_affine(image, &x, ...)) {
  handle_my_error();
  g_object_unref (image);
  return -1;
}
g_object_unref (image);
image = x;
```

Or use an unref pool, like `_local_array()`.

Summary: 

1. Make a benchmark and test `composite`, it might be good enough for you.
2. If it becomes too slow with that many inputs, consider improving the search it uses.
2a. If it's still too slow, then try your `arraymontage` idea.
3. Check your platform limits for file descriptors.
4. libvips memory use scales with output image width and with number of threads. For very wide images, you may need to limit concurrency.

**Comment by drrudolph** [comment]
Hi,
thank you very much for the comprehensive advice.
I seem to have overlooked vips_composite in the documentation somehow.
The file descriptor limit problem did in fact occur early on.

So far I tried three different montage methods:
a) simple for loop over tiles and montaging with a vips_draw_image() call each
b) same loop with vips_insert()
c) create arrays with tiles and coordinates and use vips_combine()

running on 10 cores on a machine with 64GB of RAM with latest git version of vips:
a) (vips_draw loop): memory: high-water mark 33.85 GB, 1:02
b) (vips_insert loop): memory: high-water mark 53.55 GB, 1:12
c) (vips_combine): terminated after 25 minutes (OOM)

Some additional info (example image): 
2000 tiles (~3100x2700), final image size  123143 x 156208, tiff file size ~18GB.
All tiles are in single-band UCHAR .bmp format which leads to imagemagick creating 150GB of temporary files in /tmp/, so I also tried using an external bmp library and vips_rawload while skipping the bmp-header, both are faster.

I would be happy enough with using a loop and draw_image, which is quite fast and saving to a file works fine.
However when I try to process the image further after montaging using vips_affine() or vips_mapim() the mem usage explodes and uses more than is available. I also tried to run this on a node with 370GB of RAM and got the same OOM error, I seem use vips in an inefficient way (there was a post about draw_image being problematic for the cache).
I also get the same problem using vips_combine. I tried lowering the concurrency as you suggested, but the program is still OOM-killed.

At the moment I can only think of saving the image to disk and then reloading it, which I would like to avoid if possible (speed is more important in our case than RAM usage).
I would be glad for some possible hints about using affine or mapim on large images.

**Comment by jcupitt** [comment]
We need to make a tiny benchmark program for testing --- I don't think you can debug CPU and memory problems without something concrete to measure.

I would use python and paste it here, eg.:

```python
#!/usr/bin/python3
  
import sys
import random
import pyvips

bg = pyvips.Image.black(50000, 50000)

fg = [pyvips.Image.new_from_file(filename, access='sequential')
      for filename in sys.argv[2:]]
xes = [random.randint(0, bg.width) for i in range(len(fg))]
yes = [random.randint(0, bg.height) for i in range(len(fg))]

out = bg.composite(fg, "over", x=xes, y=yes)

out.write_to_file(sys.argv[1])
```

So that's making a 60k x 60k pixel image (about 10gb, or 3 gigapixels) and pasting all of the command-line args in at random positions.

I can run it like this:

```
$ for i in {1..1000}; do cp ~/pics/k2.jpg $i.jpg; done
$ /usr/bin/time -f %M:%e ./composite-many.py x.jpg *.jpg
3196084:126.49
```

Where `k2.jpg` is a 1450 x 2048 pixel rgb jpg image. On this modest two core four thread 2015 laptop, that's 3gb of memory and 2m of processing.

If I remove the `access=` it'll have to load all of the source images into memory before starting work, and memuse rises to 7gb and CPU to 300s.

If I run with `access=` and turn concurrency down, it's 2.2gb and 220s, so lower peak memory, but obviously slower, since it can't decode in parallel.

You'd need to see how resource use scaled for your case.

**Comment by drrudolph** [comment]
I tried your python script with some minor changes to get closer to what I am trying to use it for.
I also used 2500 subimages (3128x2724).
As I need to rotate/shift the assembled image I can not use access="sequential".

```
#!/usr/bin/python3
import sys
import random
import pyvips

bg = pyvips.Image.black(150000, 150000)

fg = [pyvips.Image.new_from_file(filename, access="random") for filename in sys.argv[2:]]

xes = [random.randint(0, bg.width) for i in range(len(fg))]
yes = [random.randint(0, bg.height) for i in range(len(fg))]

out = bg.composite(fg, "over", x=xes, y=yes)

out = out.flatten()

rotated = out.rotate(32)
rotated.write_to_file(sys.argv[1])
```

If I load the tiles from .tif files, everything runs fine.
```
        User time (seconds): 3360.49
	System time (seconds): 64.91
	Percent of CPU this job got: 1084%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 5:15.74
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 26535456
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 6693754
	Voluntary context switches: 10822406
	Involuntary context switches: 4991071
	Swaps: 0
	File system inputs: 13817848
	File system outputs: 83443144
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 
```

If I use the same tiles but in .bmp format, the /tmp/ directory gets filled with up with magick-* files until there is no space left, the script stops after 10 minutes with:
```
Traceback (most recent call last):
  File "/home/daniel/vips_composite_test.py", line 9, in <module>
    fg = [pyvips.Image.new_from_file(filename, access="random") for filename in sys.argv[2:]]
  File "/home/daniel/vips_composite_test.py", line 9, in <listcomp>
    fg = [pyvips.Image.new_from_file(filename, access="random") for filename in sys.argv[2:]]
  File "/home/daniel/.local/lib/python3.7/site-packages/pyvips/vimage.py", line 217, in new_from_file
    raise Error('unable to load from file {0}'.format(vips_filename))
pyvips.error.Error: unable to load from file b'864.bmp'
  VipsForeignLoad: "864.bmp" is not a known file format
...[list of images]...
	User time (seconds): 1022.34
	System time (seconds): 375.85
	Percent of CPU this job got: 223%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 10:25.65
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 25279728
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 57415
	Minor (reclaiming a frame) page faults: 110977636
	Voluntary context switches: 187000
	Involuntary context switches: 1728395
	Swaps: 0
	File system inputs: 39101264
	File system outputs: 689415384
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 1
```
I suppose that is because imagemagick is unable to create any more temporary files. This seems to be the same problem I ran into using C. If I limit the disk cache in /etc/ImageMagick-6/policy.xml so that the disk is not completely filled, I get the same result, only sooner.

In summary, it does not seem to be a problem with vips directly, but with ImageMagick caching and any additional crashes in my program are my own fault ...

I guess I could try to write a vips_load_bmp or load .bmp files with vips_rawload.

Thanks for your assistance.

**Comment by jcupitt** [comment]
It depends a bit on the TIFF -- you should find that tiled TIFFs work best. They support random access, so libvips can load pixels from them on demand with no preprocessing.

You'll find a implementation based on `rawload` will only work for a subset of BMPs, unfortunately, since BMP has quite a few variants. It might be easier to have a separate BMP -> something else conversion step before processing.
